ioutilx/copy.go

```diff
@@ -0,0 +1,409 @@
+package ioutilx
+
+// Note that these functions may not be portable to all systems. Specifically,
+// none of this is tested on Windows.
+//
+// This code is based on: https://github.com/termie/go-shutil
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"syscall"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/teamwork/utils/sliceutil"
+)
+
+// SameFileError is used when the source and destination file are the same file.
+type SameFileError struct {
+	Src string
+	Dst string
+}
+
+func (e SameFileError) Error() string {
+	return fmt.Sprintf("%v and %v are the same file", e.Src, e.Dst)
+}
+
+// AlreadyExistsError is used when the destination already exists.
+type AlreadyExistsError struct {
+	Dst string
+}
+
+func (e AlreadyExistsError) Error() string {
+	return fmt.Sprintf("%s already exists", e.Dst)
+}
+
+// NotADirectoryError is used when attempting to copy a directory tree that is
+// not a directory.
+type NotADirectoryError struct {
+	Src string
+}
+
+func (e NotADirectoryError) Error() string {
+	return fmt.Sprintf("%s is not a directory", e.Src)
+}
+
+// SpecialFileError is used when the source or destination file is a special
+// file, and not something we can operate on.
+type SpecialFileError struct {
+	FileInfo os.FileInfo
+}
+
+func (e SpecialFileError) Error() string {
+	var mode string
+	switch {
+	case (e.FileInfo.Mode() & os.ModeDevice) == os.ModeDevice:
+		mode = "device file"
+	case (e.FileInfo.Mode() & os.ModeNamedPipe) == os.ModeNamedPipe:
+		mode = "named pipe"
+	case (e.FileInfo.Mode() & os.ModeSocket) == os.ModeSocket:
+		mode = "domain socket"
+	case (e.FileInfo.Mode() & os.ModeCharDevice) == os.ModeCharDevice:
+		mode = "character device"
+	default:
+		panic("this should never happen")
+	}
+
+	return fmt.Sprintf("%v is not a regular file but a %v",
+		e.FileInfo.Name(), mode)
+}
+
+// IsSameFile reports if two files refer to the same file object. It will return
+// a SameFileError if it is.
+//
+// It is not an error if one of the two files doesn't exist.
+func IsSameFile(src string, dst string) error {
+	srcInfo, err := os.Stat(src)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil
+		}
+		return errors.WithStack(err)
+	}
+
+	dstInfo, err := os.Stat(dst)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil
+		}
+		return errors.WithStack(err)
+	}
+
+	if os.SameFile(srcInfo, dstInfo) {
+		return &SameFileError{Src: src, Dst: dst}
```

Comment:
A bit unexpected to ask a question and get an error to indicate a positive answer. Maybe it should resemble more `os.SameFile`